File: app/src/main/java/com/example/myapplication/weight/RadarView.java
package com.example.myapplication.weight;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.SweepGradient;
import android.graphics.drawable.ShapeDrawable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import androidx.annotation.Nullable;
//TODO 雷达
public class RadarView extends View {
    public static final String TAG = "RadarView";

    public RadarView(Context context) {
        super(context);
        init();
    }

    public RadarView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public RadarView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    //园的画笔和雷达的画笔
    Paint mPaintCircle, mPaintRader;
    //当前雷达度数
    int nowDegrees = 0;
    //雷达素的
    int speed = 5;
    //渐变 梯度渲染
    SweepGradient sweepGradient;
    //半径
    int radius;
    //宽
    int width;
    //高
    int height;

    Matrix matrix = new Matrix();

    float[] radiusList = new float[]{0.1f, 0.3f, 0.5f, 0.7f, 0.9f};

    public void init() {
        mPaintCircle = new Paint();
        mPaintCircle.setStyle(Paint.Style.STROKE);
        mPaintCircle.setAntiAlias(true);
        mPaintCircle.setStrokeWidth(1);
        mPaintCircle.setColor(Color.parseColor("#B0C4DE"));

        mPaintRader = new Paint();
        mPaintRader.setAntiAlias(true);
        mPaintRader.setStyle(Paint.Style.FILL_AND_STROKE);

        post(new Runnable() {
            @Override
            public void run() {
                nowDegrees = (nowDegrees + speed) % 360;
                matrix.setRotate(nowDegrees, width / 2, height / 2);
                invalidate();
                postDelayed(this, 50);
            }
        });
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        Log.i(TAG, "onMeasure");
        height = getMeasuredHeight();
        width = getMeasuredWidth();
        radius = Math.min(height, width) / 2;
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        Log.i(TAG, "onSizeChanged");
        sweepGradient = new SweepGradient(width / 2, height / 2, new int[]{Color.TRANSPARENT, Color.parseColor("#84B5CA")}, null);
        mPaintRader.setShader(sweepGradient);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        Log.i(TAG, "onDraw");
        for (float r : radiusList) {
            canvas.drawCircle(width / 2, height / 2, radius * r, mPaintCircle);
        }

        canvas.save();
        canvas.concat(matrix);
        canvas.drawCircle(width / 2, height / 2, radius, mPaintRader);
        canvas.restore();
//        sweepGradient.setLocalMatrix(matrix);
//        canvas.drawCircle(width/2,height/2,radius,mPaintRader);


    }
}

File: app/src/main/java/com/example/myapplication/gitflow/weight/matrix/RadarView.kt
package com.example.myapplication.gitflow.weight.matrix

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import kotlin.math.min

/**
 * 雷达
 */
class RadarView : View {
    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)

    lateinit var sweepGradient: SweepGradient

    //TODO 雷达直径比例
    var radiusList = floatArrayOf(0F, 0.2F, 0.4F, 0.6F, 0.8F)
    var _paint = Paint()
    var _paintBg = Paint()
    var _matrix: Matrix = Matrix()

    //TODO 宽高
    var _width = 0
    var _height = 0
    var _radius = 0

    //TODO 旋转角度
    var nowDegrees = 0
    var speed = 10

    init {
        _paint.apply {
            color = Color.BLACK
            style = Paint.Style.STROKE
            isAntiAlias = true
            strokeWidth = 1F
        }
        _paintBg.apply {
            style = Paint.Style.FILL_AND_STROKE
            isAntiAlias = true
        }
        post(object : Runnable {
            override fun run() {
                nowDegrees = (nowDegrees + speed) % 360
                _matrix.setRotate(nowDegrees.toFloat(), (_width / 2).toFloat(), (_height / 2).toFloat())
                invalidate()
                postDelayed(this, 50)
            }
        })
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        _width = measuredWidth
        _height = measuredHeight
        _radius = min(_width, _height) / 2
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)

        sweepGradient = SweepGradient((_width / 2).toFloat(), (_height / 2).toFloat(), intArrayOf(Color.TRANSPARENT, Color.GRAY), floatArrayOf(0F, 1F))
        _paintBg.shader = sweepGradient
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        canvas?.drawColor(Color.BLACK)
        radiusList.forEach {
            canvas?.drawCircle((_width / 2).toFloat(), (_height / 2).toFloat(), it * _radius.toFloat(), _paint)
        }
        canvas?.save()
        sweepGradient.setLocalMatrix(_matrix)
        canvas?.drawCircle((_width / 2).toFloat(), (_height / 2).toFloat(), _radius.toFloat(), _paintBg)
        canvas?.restore()
    }
}